<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Analysis of single cell data with pagoda</title>

<script type="text/javascript">
window.onload = function() {
  var imgs = document.getElementsByTagName('img'), i, img;
  for (i = 0; i < imgs.length; i++) {
    img = imgs[i];
    // center an image if it is the only element of its parent
    if (img.parentElement.childElementCount === 1)
      img.parentElement.style.textAlign = 'center';
  }
};
</script>





<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 13px;
}

body {
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 20px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 {
   font-size:2.2em;
}

h2 {
   font-size:1.8em;
}

h3 {
   font-size:1.4em;
}

h4 {
   font-size:1.0em;
}

h5 {
   font-size:0.9em;
}

h6 {
   font-size:0.8em;
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre, img {
  max-width: 100%;
}
pre {
  overflow-x: auto;
}
pre code {
   display: block; padding: 0.5em;
}

code {
  font-size: 92%;
  border: 1px solid #ccc;
}

code[class] {
  background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * {
      background: transparent !important;
      color: black !important;
      filter:none !important;
      -ms-filter: none !important;
   }

   body {
      font-size:12pt;
      max-width:100%;
   }

   a, a:visited {
      text-decoration: underline;
   }

   hr {
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote {
      padding-right: 1em;
      page-break-inside: avoid;
   }

   tr, img {
      page-break-inside: avoid;
   }

   img {
      max-width: 100% !important;
   }

   @page :left {
      margin: 15mm 20mm 15mm 10mm;
   }

   @page :right {
      margin: 15mm 10mm 15mm 20mm;
   }

   p, h2, h3 {
      orphans: 3; widows: 3;
   }

   h2, h3 {
      page-break-after: avoid;
   }
}
</style>



</head>

<body>
<h1>Analysis of single cell data with pagoda</h1>

<pre><code class="R">library(&#39;pagoda2&#39;)
library(Matrix)

##############################
# Data preparation
##############################

# Read data from your file, rows as genes colums as cells
myCountMatrix &lt;- read.table(&#39;mySCdata.txt&#39;);

# Make the gene name uniquye
rownames(myCountMatrix) &lt;- make.unique(rownames(myCountMatrix))

# Conver the matrix to a sparse matrix
myCountMatrixSparse &lt;- Matrix(myCountMatrix, sparse = T)

# Remove the original matrix
rm(myCountMatrix); 




##############################
# Process the data
##############################

# Generate a new pagoda2 object
myPagoda2Object &lt;- Pagoda2$new(x = myCountMatrixSparse, n.cores = 4)

# Adjust the variance
myPagoda2Object$adjustVariance(plot = T, gam.k = 10)


# Calculate a PCA reduction with the number of PCs specified by nPCs
# and using only the n.odgenes overdispersed genes -- in this case 2000
myPagoda2Object$calculatePcaReduction(nPcs = 100, n.odgenes = 2.e3)

# Generate K-nearest neighbour graph
myPagoda2Object$makeKnnGraph(k = 20, type = &#39;PCA&#39;, center = T,
    weight.type = &#39;none&#39;, n.cores = 30, distance = &#39;cosine&#39;)


##############################
# Identify clusters
##############################

# Identify clusters using the infomap.community method
# on the basis of the reduction called &#39;PCA&#39; (generated above)
# Save the resulting clustering as &#39;infomap&#39;
myPagoda2Object$getKnnClusters(method = infomap.community,
                               type = &#39;PCA&#39;, name = &#39;infomap&#39;)

# Do an independent identification of clusters using the
# multilevel community algorithm again using the PCA reduction
# and save it as &#39;multilevel&#39;. This does not overwrite the 
# previous clustering
myPagoda2Object$getKnnClusters(method = multilevel.community,
                               type = &#39;PCA&#39;, name=&#39;multilevel&#39;)

# Do yet another clustering
myPagoda2Object$getKnnClusters(method = walktrap.community,
                               type = &#39;PCA&#39;, name=&#39;walktrap&#39;)



##############################
# Generate embeddings
##############################
# Generate an embedding with largeVis on the basis of the PCA reduction
M &lt;- 30
myPagoda2Object$getEmbedding(
  type = &#39;PCA&#39;,
  embeddingType = &#39;largeVis&#39;,
  M = M,
  perplexity = 30,
  gamma = 1 / M,
  alpha = 1
)

# Generate an embedding with tSNE on the basis of the PCA
# reduction
myPagoda2Object$getEmbedding(type = &#39;PCA&#39;, embeddingType = &#39;tSNE&#39;)

##############################
# Plot the generated embeddings
##############################
myPagoda2Object$plotEmbedding(type = &#39;PCA&#39;,
                              embedding = &#39;largeVis&#39;,
                              mark.clusters = T,
                              clusterType = &#39;infomap&#39;)


myPagoda2Object$plotEmbedding(type = &#39;PCA&#39;,
                              embedding = &#39;largeVis&#39;,
                              mark.clusters = T,
                              clusterType = &#39;multilevel&#39;)
myPagoda2Object$plotEmbedding(type = &#39;PCA&#39;,
                              embedding = &#39;largeVis&#39;,
                              mark.clusters = T,
                              clusterType = &#39;walktrap&#39;)


myPagoda2Object$plotEmbedding(type = &#39;PCA&#39;,
                              clusterType = &#39;infomap&#39;,
                              embeddingType = &#39;tSNE&#39;,
                              mark.clusters = T)

##############################
# Differential Gene expression
##############################

# Calculate the differential gene expression of each cluster
# defined in multilevel clustring on the basis of the PCA reduction
myPagoda2Object$getDifferentialGenes(type=&#39;PCA&#39;,clusterType = &#39;multilevel&#39;)

# Plot a differential expression heatmap using the differentially expressed genes
# above
myPagoda2Object$plotDiffGeneHeatmap(type = &#39;PCA&#39;, clusterType = &#39;multilevel&#39;)



# Pathway overdispersion -- required for web
go.env &lt;- p2.generate.human.go(myPagoda2Object)
myPagoda2Object$testPathwayOverdispersion(setenv = go.env, verbose = T, correlation.distance.threshold = 0.9, 
                                          recalculate.pca = F,
                                          min.pathway.size = 100, max.pathway.size = 1000)




################################
# Generate the web application
################################

# Generate GO genesets for the web app
myGeneNames &lt;- colnames(myPagoda2Object$counts)
goSets &lt;- p2.generate.human.go.web(myGeneNames)

# Generate differental expression between each cluster and everything else
# Load these clusters as pre-defined gene sets with the given prefix
deSets &lt;- get.de.geneset(myPagoda2Object, groups = myPagoda2Object$clusters$PCA[[1]], prefix = &#39;de_&#39;)

# Merge Genesets
geneSets &lt;- c(goSets, deSets)

# Additional metadata generation
additionalMetadata &lt;- list()
additionalMetadata$altCluster &lt;- p2.metadata.from.factor(myPagoda2Object$clusters$PCA[[1]], displayname = &#39;Infomap&#39;, s = 0.7, v = 0.8,start = 0, end = 0.5)
additionalMetadata$altCluster &lt;- p2.metadata.from.factor(myPagoda2Object$clusters$PCA[[2]], displayname = &#39;Multilevel&#39;, s = 0.7, v = 0.8,start = 0, end = 0.5)

# Set the palette manually for Walktrap
a &lt;- myPagoda2Object$clusters$PCA[[3]]
library(colorRamps)
p1 &lt;- colorRamps::primary.colors(n = nlevels(a))
names(p1) &lt;- levels(a)
additionalMetadata$altCluster2 &lt;- p2.metadata.from.factor(myPagoda2Object$clusters$PCA[[3]], displayname = &#39;Walktrap&#39;, pal = 1)


# Generate the gene knn graph, which is used to find gene with similar expression patterns
# This step is required for web object serialisation with the serializeToStaticFast() function
myPagoda2Object$geneKnnbyPCA()

# Generate and display web app
myPagoda2WebObject &lt;-
  make.p2.app(
    myPagoda2Object,
    dendrogramCellGroups = myPagoda2Object$clusters$PCA[[1]],
    additionalMetadata = additionalMetadata,
    geneSets = geneSets,
    show.clusters = FALSE, # Hide the clusters that were used for the dendrogram from the metadata
  );

# For matching server env
rm(myPagoda2Object); # The original object is no longer required

# This step is optional, it is required if the app is loaded from 
# disk on another machine where the installation of the R package
# is on another directory
# myPagoda2WebObject$updateRookRoot();

# Show the app
show.app(app=myPagoda2WebObject,name=&#39;p2Sample1&#39;)

# Optionally serialise to a static binary file
# tmp/ directory must exist and be empty
# (Also requires packing utility in the utilities directory of the package to be built
# Provides error with instructions if it is not)

# Deprecated, use newer and faster method below
# myPagoda2WebObject$serialiseToStatic(text.file.directory = &#39;tmp/&#39;, binary.filename = &#39;pagodaTest.bin&#39;)


myPagoda2WebObject$serializeToStaticFast(binary.filename = &#39;pagodaTestNew.bin&#39;);

# Save the app as an RDS for deploying
saveRDS(myPagoda2WebObject, file = &#39;~/p2Sample1.rds&#39;) # This object will now support DE

</code></pre>

</body>

</html>
